### PR TITLE
refactor: task with context

### PR DESCRIPTION
**Pull request by antonioiubatti93**
Do not force users to define lambda functions based on an outer context.

BREAKING CHANGE: input (blocked) tasks take a context as argument.

### CODE REVIEW

File: octopus.go [hunk 1]
```diff
@@ -38,16 +38,19 @@ func (o *Octopus) Run(ctx context.Context) Termination {
...
 	}
 
 	for _, blockedTask := range o.blockedTasks {
 		blockedTask := blockedTask
 
 		eg.Go(func() error {
 			<-wait()
-
-			return blockedTask()
+			return blockedTask(ctx)
```
Comment (caroline105) [inline octopus.go:53]: What will happen if the context expires before the blocked tasks are launched? Those tasks will never be executed?

Comment (antonioiubatti93) [reply octopus.go:53]: For the blocked tasks, it's the opposite: whatever the event that caused the "free" tasks to terminate (either context cancellation or signal interception), the `defer done()` statement in the last goroutine will "de-block" `<-wait()` and execute the blocked tasks right after.
This code will eventually disappear, but the question was worth an answer :wink:

Comment (caroline105) [reply octopus.go:53]: But as the blocked task is itself terminated when the context is canceled, will it ever be executed after the `<-wait()` is deblocked by a context cancellation?

**Review by caroline105 — APPROVED** [review]

**Comment by antonioiubatti93** [comment]
This PR will be merged after #13 .

**Comment by antonioiubatti93** [comment]
Opening a refactoring PR.

**Pull request closed without merging**